net/xia/ppal_ether/main.c

```diff
@@ -0,0 +1,882 @@
+	u32 seq = cb->nlh->nlmsg_seq;
+	struct rtmsg *rtm;
+	struct fib_xid_ether_main *mether = fxid_mether(fxid);
+	struct xia_xid dst;
+	struct nlattr *ha_attr;
+	struct rtnl_xia_ether_addrs *rtha;
+	struct interface_addr *pos_ia;
+
+	nlh = nlmsg_put(skb, portid, seq, RTM_NEWROUTE, sizeof(*rtm),
+			NLM_F_MULTI);
+	if (nlh == NULL)
+		return -EMSGSIZE;
+
+	rtm = nlmsg_data(nlh);
+	rtm->rtm_family = AF_XIA;
+	rtm->rtm_dst_len = sizeof(struct xia_xid);
+	rtm->rtm_src_len = 0;
+	rtm->rtm_tos = 0; /* XIA doesn't have a tos. */
+	rtm->rtm_table = XRTABLE_MAIN_INDEX;
+	/* XXX One may want to vary here. */
+	rtm->rtm_protocol = RTPROT_UNSPEC;
+	/* XXX One may want to vary here. */
+	rtm->rtm_scope = RT_SCOPE_UNIVERSE;
+	rtm->rtm_type = RTN_UNICAST;
+	/* XXX One may want to put something here, like RTM_F_CLONED. */
+	rtm->rtm_flags = 0;
+
+	dst.xid_type = xtbl->fxt_ppal_type;
+	memmove(dst.xid_id, fxid->fx_xid, XIA_XID_MAX);
+	if (unlikely(nla_put(skb, RTA_DST, sizeof(dst), &dst)))
+		goto nla_put_failure;
+
+	/* Hardware addresses. */
+	ha_attr = nla_nest_start(skb, RTA_MULTIPATH);
+	if (!ha_attr)
+		goto nla_put_failure;
+
+	rtha = nla_reserve_nohdr(skb, sizeof(*rtha));
+	if (!rtha)
+		goto nla_put_failure;
+
+	rcu_read_lock();
+	pos_ia = rcu_dereference(mether->neigh_addr);
+
+	rtha->interface_addr_len = pos_ia->outgress_interface->addr_len;
+	memmove(rtha->interface_addr, pos_ia->ha, rtha->interface_addr_len);
+	rtha->interface_index = pos_ia->outgress_interface->ifindex;
+	
+	rcu_read_unlock();
+	
+	/* No attributes. */
+
+	/* length of rtnetlink header + attributes */
+	rtha->attr_len = nlmsg_get_pos(skb) - (void *)rtha;
+
+	nla_nest_end(skb, ha_attr);
+
+	nlmsg_end(skb, nlh);
+	return 0;
+
+nla_put_failure:
+	nlmsg_cancel(skb, nlh);
+	return -EMSGSIZE;
+}
+
+/* call using fxid_free only */
+void main_free_ether(struct fib_xid_table *xtbl, struct fib_xid *fxid)
+{
+	struct fib_xid_ether_main *mether = fxid_mether(fxid);
+
+	xdst_free_anchor(&mether->xem_anchor);
+	mether->xem_dead = true;
+	mether_finish_destroy(mether);
+}
+
+/* ETHER_FIB all table external operations */
+static const xia_ppal_all_rt_eops_t ether_all_rt_eops = {
+	[XRTABLE_LOCAL_INDEX] = {
+		.newroute = local_newroute,
+		.delroute = local_delroute,
+		.dump_fxid = local_dump_ether,
+		.free_fxid = local_free_ether,
+	},
+
+	[XRTABLE_MAIN_INDEX] = {
+		.newroute = main_newroute,
+		.delroute = main_delroute,
+		.dump_fxid = main_dump_ether,
+		.free_fxid = main_free_ether,
+	},
+};
+
+/* routing process per principal struct */
+
+static inline struct interface_addr *xdst_naddr(struct xip_dst *xdst)
+{
+	return xdst->info;
+}
+
+static int main_input_input(struct sk_buff *skb)
+{
+	struct xiphdr *xiph;
+	struct xip_dst *xdst;
+	struct interface_addr *naddr;
+
+	/* XXX We should test that forwarding is enable per struct net.
+	 * See example in net/ipv6/ip6_output.c:ip6_forward.
+	 * TODO:ask why needed to test coz already path has been decided
```

Comment:
Why do we leave a ``TODO`` tag here? Do we still need to add more functionality here?